From 32811892d277a98f052496222362460a043ca6c9 Mon Sep 17 00:00:00 2001
From: Christoph Hellwig <hch@lst.de>
Date: Wed, 1 Sep 2021 13:38:32 +0200
Subject: [PATCH 045/905] [Backport] md: extend disks_mutex coverage

mainline inclusion
from mainline-v5.16-rc1
commit 94f3cd7d832c28681f1dea54b4dd8606e5e2bc75
category: bugfix
bugzilla: 188015

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=94f3cd7d832c28681f1dea54b4dd8606e5e2bc75

--------------------------------

disks_mutex is intended to serialize md_alloc.  Extended it to also cover
the kobject_uevent call and getting the sysfs dirent to help reducing
error handling complexity.

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Song Liu <songliubraving@fb.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>

Conflict:
	drivers/md/md.c

Signed-off-by: Li Nan <linan122@huawei.com>
---
 drivers/md/md.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/md/md.c b/drivers/md/md.c
index bd255dcc6e37..05bc734d59d5 100644
--- a/drivers/md/md.c
+++ b/drivers/md/md.c
@@ -5743,12 +5743,12 @@ static int md_alloc(dev_t dev, char *name)
 	    sysfs_create_group(&mddev->kobj, &md_bitmap_group))
 		pr_debug("pointless warning\n");
  abort:
-	mutex_unlock(&disks_mutex);
 	if (!error && mddev->kobj.sd) {
 		kobject_uevent(&mddev->kobj, KOBJ_ADD);
 		mddev->sysfs_state = sysfs_get_dirent_safe(mddev->kobj.sd, "array_state");
 		mddev->sysfs_level = sysfs_get_dirent_safe(mddev->kobj.sd, "level");
 	}
+	mutex_unlock(&disks_mutex);
 	mddev_put(mddev);
 	return error;
 }
-- 
2.17.1

